zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -748,82 +758,115 @@ pub(crate) fn update_chain_tip<P: consensus::Parameters>(
-    let shard_entry = shard_start_height
+    let tip_shard_entry = shard_start_height
         .filter(|h| h < &chain_end)
         .map(|h| ScanRange::from_parts(h..chain_end, ScanPriority::ChainTip));
 
-    // Create scanning ranges to either validate potentially invalid blocks at the wallet's view
-    // of the chain tip, or connect the prior tip to the new tip.
-    let tip_entry = prior_tip.map(|prior_tip| {
-        // If we don't have shard metadata, this means we're doing linear scanning, so create a
-        // scan range from the prior tip to the current tip with `Historic` priority.
-        if shard_entry.is_none() {
-            ScanRange::from_parts(prior_tip..chain_end, ScanPriority::Historic)
-        } else {
-            // Determine the height to which we expect blocks retrieved from the block source to be stable
-            // and not subject to being reorg'ed.
-            let stable_height = new_tip.saturating_sub(PRUNING_DEPTH);
-
-            // If the wallet's prior tip is above the stable height, prioritize the range between
-            // it and the new tip as `ChainTip`. Otherwise, prioritize the `VERIFY_LOOKAHEAD`
-            // blocks above the wallet's prior tip as `Verify`. Since `scan_cached_blocks`
-            // retrieves the metadata for the block being connected to, the connectivity to the
-            // prior tip will always be checked. Since `Verify` ranges have maximum priority, even
-            // if the block source begins downloading blocks from the shard scan range (which ends
-            // at the stable height) the scanner should not attempt to scan those blocks until the
-            // tip range has been completely checked and any required rewinds have been performed.
-            if prior_tip >= stable_height {
-                // This may overlap the `shard_entry` range and if so will be coalesced with it.
-                ScanRange::from_parts(prior_tip..chain_end, ScanPriority::ChainTip)
+    // Create scan ranges to either validate potentially invalid blocks at the wallet's
+    // view of the chain tip, or connect the prior tip to the new tip.
+    let tip_entry = max_scanned.map_or_else(
+        || {
+            // No blocks have been scanned, so we need to anchor the start of the new scan
+            // range to something else.
+            wallet_birthday.map_or_else(
+                // We don't have a wallet birthday, which means we have no accounts yet.
+                // We can therefore ignore all blocks up to the chain tip.
+                || ScanRange::from_parts(sapling_activation..chain_end, ScanPriority::Ignored),
+                // We have a wallet birthday, so mark all blocks between that and the
+                // chain tip as `Historic` (performing wallet recovery).
+                |wallet_birthday| {
+                    ScanRange::from_parts(wallet_birthday..chain_end, ScanPriority::Historic)
+                },
```

Comment:
Shouldn't we also be marking blocks between Sapling activation and the wallet birthday as `Ignored` when we have a birthday?